### PR TITLE
feat(shorebird_cli): add IpaDiffer to get changes between to ipa files

### PR DESCRIPTION
**Pull request by bryanoltman**
## Description

Adds `IpaDiffer` to read .ipa files and find differences between them.

This also refactors `ArchiveDiffer` and its subclasses to improve code sharing and removes some obsolete tests for `AarDiffer` and `AabDiffer`.

Part of https://github.com/shorebirdtech/shorebird/issues/1001

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

**Comment by eseidel** [comment]
Tools that may be of help:

https://gist.github.com/mokemokechicken/bc33b2b19a0dd3a76084
https://github.com/alexzielenski/optool
https://www.manpagez.com/man/1/codesign/
codesign --remove-signature

File: packages/shorebird_cli/lib/src/archive_analysis/ipa_differ.dart [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+import 'package:shorebird_cli/src/archive_analysis/archive_differ.dart';
+import 'package:shorebird_cli/src/archive_analysis/file_set_diff.dart';
+
+/// Finds differences between two IPAs.
+///
+/// Asset changes will be in the `Assets.car` file (which is a combination of
+/// the `.xcasset` catalogs in the Xcode project) and the `flutter_assets`
+/// directory.
+///
+/// Native changes will appear in the Runner.app/Runner executable.
```
Comment (eseidel) [inline packages/shorebird_cli/lib/src/archive_analysis/ipa_differ.dart]: And any .framework included.

File: packages/shorebird_cli/lib/src/archive_analysis/ipa_differ.dart [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        newPathHashes: _fileHashes(File(newArchivePath)),
+      );
+
+  @override
+  bool containsPotentiallyBreakingAssetDiffs(FileSetDiff fileSetDiff) =>
+      assetsFileSetDiff(fileSetDiff).isNotEmpty;
+
+  @override
+  bool containsPotentiallyBreakingNativeDiffs(FileSetDiff fileSetDiff) {
+    // Because the mach-o binaries are signed, they will always have different
```
Comment (eseidel) [inline packages/shorebird_cli/lib/src/archive_analysis/ipa_differ.dart]: Specifically that they're signed with an expiry date included, I think.

File: packages/shorebird_cli/lib/src/archive_analysis/ipa_differ.dart [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      filePath.endsWith('App.framework/App');
+
+  @override
+  bool isNativeFilePath(String filePath) => appRegex.hasMatch(filePath);
+
+  PathHashes _fileHashes(File ipa) {
+    final zipDirectory = ZipDirectory.read(InputFileStream(ipa.path));
+    return {
+      for (final file in zipDirectory.fileHeaders)
+        file.filename: file.crc32!.toString()
```
Comment (eseidel) [inline packages/shorebird_cli/lib/src/archive_analysis/ipa_differ.dart:71]: I might add a comment explaining what this is and why it's OK.

// Zip files contain an (optional) crc32 checksum for a file.  IPAs seem to always
// include this for files, so a quick way for us to tell if file contents differ is if
// their checksums differ.

**Review by eseidel — COMMENTED** [review]
overall lgtm, will let felix press the actual button.

File: packages/shorebird_cli/lib/src/archive_analysis/ipa_differ.dart (hunk 3 above)
Comment (bryanoltman) [reply packages/shorebird_cli/lib/src/archive_analysis/ipa_differ.dart:71]: Done.

File: packages/shorebird_cli/lib/src/archive_analysis/android_archive_differ.dart [hunk 4]
```diff
@@ -22,33 +14,19 @@ abstract class AndroidArchiveDiffer extends ArchiveDiffer {
     }
 
     return assetsDiff.changedPaths
```
Comment (eseidel) [inline packages/shorebird_cli/lib/src/archive_analysis/android_archive_differ.dart:16]: I guess this is also assetsDiff.changePaths.map(p.basename).toSet().difference(ArchiveDiffer.assetFileNamesToIgnore.toSet()).isEmpty(), but that's no shorter. 🤣

File: packages/shorebird_cli/lib/src/archive_analysis/archive_differ.dart [hunk 5]
```diff
@@ -1,19 +1,75 @@
...
-  bool containsPotentiallyBreakingAssetDiffs(FileSetDiff fileSetDiff);
+  bool containsPotentiallyBreakingAssetDiffs(FileSetDiff fileSetDiff) {
+    final assetsDiff = assetsFileSetDiff(fileSetDiff);
+
+    // If assets were added, we need to warn the user about asset differences.
+    if (assetsDiff.addedPaths.isNotEmpty) {
+      return true;
+    }
+
+    return assetsDiff.changedPaths
```
Comment (eseidel) [inline packages/shorebird_cli/lib/src/archive_analysis/archive_differ.dart:31]: Given you've repeated this complex whereNot twice, I might have written this:

return removeIgnoredFiles(assetsDiff.changedPaths).isEmpty() or similar.

**Review by eseidel — APPROVED** [review]

File: packages/shorebird_cli/lib/src/archive_analysis/archive_differ.dart (hunk 5 above)
Comment (bryanoltman) [reply packages/shorebird_cli/lib/src/archive_analysis/archive_differ.dart:31]: This is addressed in the followup at https://github.com/shorebirdtech/shorebird/pull/1011

**Pull request merged**